### DISCUSSION TITLE
How to focus on multiple selectors?

### DISCUSSION THREAD

**Discussion by lamualfa** [pmndrs/jotai#1331] [Q&A] [answered]
The SQL database restricts the user to use nested data structures using the object. So the only solution is to group the data by providing the group name as the prefix of the column name. But the problem with this approach is you'll have a single big flat object that causes re-render anytime when an object's field changes. 

Here I want to create another atom to watch only a specific field that a component needs. But Jotai's documentation only shows the example to watch only one field.  Any idea to implementing the same thing with the multiple fields or selectors?

My current idea. Haven't tried it. But does it really work and is the best practice to solve the problem?

```js
import { atom } from 'jotai'
import { selectAtom } from 'jotai/utils'

// raw data from database
const pageAtom = atom({
  meta_title: '',
  meta_description: '',
  meta_keywords: [],

  section_1_title: '',
  section_1_description: '',

  section_2_title: '',
  section_2_description: '',
  section_2_slides: [],
})

// filter only what is needed and remove the prefix name
const metaAtom = selectAtom(pageAtom, (v) => ({
  title: v.meta_title,
  description: v.meta_description,
  keywords: v.meta_keywords,
}))

// filter only what is needed and remove the prefix name
const section1Atom = selectAtom(pageAtom, (v) => ({
  title: v.section_1_title,
  description: v.section_1_description,
}))

// filter only what is needed and remove the prefix name
const section2Atom = selectAtom(pageAtom, (v) => ({
  title: v.section_2_title,
  description: v.section_2_description,
  slides: v.section_2_slides,
}))
```

**Comment by dai-shi** [comment]
Although a bit complicated, here's one solution:
```js
const metaTitleAtom = atom((get) => get(pageAtom).meta_title)
const metaDescriptionAtom = atom((get) => get(pageAtom).meta_description)
const metaKeywordsAtom = atom((get) => get(pageAtom).meta_keywords)
const metaAtom = atom((get) =>({
  title: get(metaTitleAtom),
  description: get(metaDescription),
  keywords: get(metaKeywords),
}))
```

You could also use `selectAtom` and shallowEqual for the second argument.

**Reply by lamualfa** [reply]
If we ignore the removal of the prefix, can we simplify it to this?

```js
const metaAtom = atom((get) => ({
  meta_title: get(pageAtom).meta_title,
  meta_description: get(pageAtom).meta_description,
  meta_keywords: get(pageAtom).meta_keywords,
}))
```

<hr/>

> You could also use `selectAtom` and shallowEqual for the second argument.

Do you mean like this?

```js
const metaAtom = selectAtom(
  pageAtom,
  (v) => ({
    title: v.meta_title,
    description: v.meta_description,
    keywords: v.meta_keywords,
  }),
  shallowEqual
)
```
If yes, can I use [fast-deep-equal](https://www.npmjs.com/package/fast-deep-equal) as the `shallowEqual` function? 

<hr/>

If those solutions will not work, will zustand solve the problem?

```js
import create from 'zustand'

const usePageStore = create(() => ({
  meta_title: '',
  meta_description: '',
  meta_keywords: [],

  section_1_title: '',
  section_1_description: '',

  section_2_title: '',
  section_2_description: '',
  section_2_slides: [],
}))

function useMeta() {
  return usePageStore((state) => ({
    title: state.meta_title,
    description: state.meta_description,
    keywords: state.meta_keywords,
  }))
}
```

**Comment by dai-shi** [comment]
https://github.com/pmndrs/jotai/discussions/1331#discussioncomment-3325635

> If we ignore the removal of the prefix, can we simplify it to this?
> 
> ```js
> const metaAtom = atom((get) => ({
>   meta_title: get(pageAtom).meta_title,
>   meta_description: get(pageAtom).meta_description,
>   meta_keywords: get(pageAtom).meta_keywords,
> }))
> ```
> 

This doesn't work as you want. It will create a new object value whenever `pageAtom` changes regardless of `meta_` ones or not.

**Comment by dai-shi** [comment, marked as answer]
https://github.com/pmndrs/jotai/discussions/1331#discussioncomment-3325635

> > You could also use `selectAtom` and shallowEqual for the second argument.
> 
> Do you mean like this?
> 
> ```js
> const metaAtom = selectAtom(
>   pageAtom,
>   (v) => ({
>     title: v.meta_title,
>     description: v.meta_description,
>     keywords: v.meta_keywords,
>   }),
>   shallowEqual
> )
> ```

Yes.

> If yes, can I use [fast-deep-equal](https://www.npmjs.com/package/fast-deep-equal) as the `shallowEqual` function?

No, deepEqual and shallowEqual are different. https://www.npmjs.com/search?q=shallow%20equal
You could use deepEqual if you want.

**Reply by lamualfa** [reply]
shallowEqual should be more efficient than deepEqual for those needs, right?

**Reply by dai-shi** [reply]
Yes.

**Reply by lamualfa** [reply]
Thank you.

**Comment by dai-shi** [comment]
https://github.com/pmndrs/jotai/discussions/1331#discussioncomment-3325635

> If those solutions will not work, will zustand solve the problem?
> 
> ```js
> import create from 'zustand'
> 
> const usePageStore = create(() => ({
>   meta_title: '',
>   meta_description: '',
>   meta_keywords: [],
> 
>   section_1_title: '',
>   section_1_description: '',
> 
>   section_2_title: '',
>   section_2_description: '',
>   section_2_slides: [],
> }))
> 
> function useMeta() {
>   return usePageStore((state) => ({
>     title: state.meta_title,
>     description: state.meta_description,
>     keywords: state.meta_keywords,
>   }))
> }
> ```

No, it's the same thing. Again, you need shallowEqual. zustand and jotai are both based on immutable state model.
